### ISSUE TITLE
RayCast3D "Exclude Parent" option is ignored

### ISSUE THREAD

**Issue by krazyjakee** [godot-jolt/godot-jolt#273]
If you have a CharacterBody3D with a capsule collider and a Raycast3d as a child, the Raycast collides with the parent character.
In Godots physics system, the raycast does not collide.

This isn't a high priority and easily worked around but a notable difference that may have affects elsewhere.

**Comment by krazyjakee** [comment]
Actually... don't dig too deep on this. Although there is a difference between physics engines, it may not be the fault of this option. I need to create a minimal reproduction and update this issue.

**Comment by mihe** [comment]
I just gave it a quick try, and from what I can tell it seems to work. When "Exclude Parent" is enabled it doesn't report any hits on my character body, and when it's disabled it does.

A minimal reproduction project would be much appreciated.

**Comment by mihe** [comment]
Interesting! I suppose their documentation meant to say "first physics body parent" or something. This seems like a bug on my part, since you could have any number of useful nodes inbetween the `RayCast3D` and its `PhysicsBody3D` parent.

I'll dig into it a bit later.

**Comment by mihe** [comment]
So this became a bit of a dilemma.

This problem has nothing to do with "Exclude Parent", as it turns out. Instead the problem is with one of the ray-cast parameters, called `hit_back_faces`, which isn't exposed as a property on `RayCast3D`. The Godot documentation [describes it](https://docs.godotengine.org/en/4.0/classes/class_physicsrayqueryparameters3d.html#class-physicsrayqueryparameters3d-property-hit-back-faces) as "the query will hit back faces with concave polygon shapes [...] or heightmap shapes".

Jolt does have this parameter as well, so I had simply mapped one to the other, but apparently Jolt's version of it not only hits the back-faces of concave polygons and height maps, but also hits the back-faces of convex shapes like spheres, boxes, capsules, etc.

So what you're seeing is simply the "Exclude Parent" property not working because your capsule is not the immediate parent of your `RayCast3D`, which makes the ray-cast just function as normal and thereby hit whatever it hits. Because Godot doesn't do back-face collision on convex shapes you get no hits, and because Jolt does you instead hit the back of the capsule.

So now I have to choose whether I strip functionality from Jolt (which is easy enough) in order to align with Godot and its documentation, or if I diverge and commit to this difference in behavior. Frankly I struggle to see the utility in limiting back-face hits to only concave polygons and height maps.

I need to think about this for a bit.

**Comment by mihe** [comment]
I'm gonna go ahead and just commit to this difference in behavior. It can always be changed later if this turns out be annoying in some way.

Thanks for reporting it!

**Comment by Regrad** [comment]
Hello. How can i disable hit_back_faces in my project? This prevents me from using Raycast normally.

**Comment by mihe** [comment]
The only way (as of right now) is to do the ray-cast manually using the `intersect_ray` method in `PhysicsDirectSpaceState3D`, as seen in [Godot's documentation](https://docs.godotengine.org/en/stable/tutorials/physics/ray-casting.html) for ray-casting.

So it would look something like this:

```gdscript
var params := PhysicsRayQueryParameters3D.new()
params.collide_with_areas = false
params.collide_with_bodies = true
params.collision_mask = 0xFFFFFFFF
params.exclude = [self]
params.hit_back_faces = false
params.hit_from_inside = false
params.from = your_starting_position
params.to = your_end_position

var hit := get_world_3d().direct_space_state.intersect_ray(params)
```

If you're doing this in `_integrate_forces` you can also fetch the `PhysicsDirectSpaceState3D` from the `PhysicsDirectBodyState3D` that gets passed into it as a parameter, but they're the same object in the end, so it doesn't make any real difference.

I'll see about making a pull request for Godot to add `hit_back_faces` as a property to the ray-cast nodes. It's a bit strange that nobody else has done it yet.

**Comment by Regrad** [comment]
Thank you. Is it possible to make a global shutdown in the settings as a temporary solution?

**Comment by mihe** [comment]
Possible, certainly, but I'd rather not go down that route.

You can always add something like the above mentioned code to a node of your own, and replace `RayCast3D` entirely. There's not much more to it really, if you disregard the editor gizmo for it.

If you're able/willing to [build from source](https://github.com/godot-jolt/godot-jolt/blob/master/docs/building.md) you can also hardcode the thing yourself here:

https://github.com/godot-jolt/godot-jolt/blob/a2d595c62b91e7dd6fbd07720973a1a676ff0b24/src/spaces/jolt_physics_direct_space_state_3d.cpp#L44-L46

**Comment by mihe** [comment]
I went ahead and created a pull request: godotengine/godot#79330.

**Comment by Regrad** [comment]
Hit from inside (off) is not work

**Comment by krazyjakee** [comment]
@Regrad this has already been answered above 

> The only way (as of right now) is to do the ray-cast manually using the `intersect_ray` method in `PhysicsDirectSpaceState3D`, as seen in [Godot's documentation](https://docs.godotengine.org/en/stable/tutorials/physics/ray-casting.html) for ray-casting.
> 
> So it would look something like this:
> 
> ```gdscript
> var params := PhysicsRayQueryParameters3D.new()
> params.collide_with_areas = false
> params.collide_with_bodies = true
> params.collision_mask = 0xFFFFFFFF
> params.exclude = [self]
> params.hit_back_faces = false
> params.hit_from_inside = false
> params.from = your_starting_position
> params.to = your_end_position
> 
> var hit := get_world_3d().direct_space_state.intersect_ray(params)
> ```
> 
> If you're doing this in `_integrate_forces` you can also fetch the `PhysicsDirectSpaceState3D` from the `PhysicsDirectBodyState3D` that gets passed into it as a parameter, but they're the same object in the end, so it doesn't make any real difference.
> 
> I'll see about making a pull request for Godot to add `hit_back_faces` as a property to the ray-cast nodes. It's a bit strange that nobody else has done it yet.

**Comment by mihe** [comment]
Also addressed in #542, in case anyone else gets tripped up by `hit_from_inside` and find themselves here.
